### ISSUE TITLE
`org.apache.iceberg.util.SnapshotUtil#snapshotIdAsOfTime` could make use of binary search if `org.apache.iceberg.Table#history` is guaranteed to be ordered

### ISSUE THREAD

**Issue by findinpath** [apache/iceberg#3891]
Currently the `SnapshotUtil` is going over all the history entries of a table (with O(n) complexity).
```
    Long snapshotId = null;
    for (HistoryEntry logEntry : table.history()) {
      if (logEntry.timestampMillis() <= timestampMillis) {
        snapshotId = logEntry.snapshotId();
      }
    }
```

In case that the `org.apache.iceberg.Table#history` would be guaranteed to be ordered, a binary search algorithm could be used instead (O(log n) complexity).

**Comment by RussellSpitzer** [comment]
Unfortunately I do not believe we are guaranteed that History is ordered (WAP commits) or Complete (History has a max length) so I do not believe we can rely on it. The only true way to view the lineage is the recursive approach being used.

**Comment by xndai** [comment]
> Unfortunately I do not believe we are guaranteed that History is ordered (WAP commits) or Complete (History has a max length) so I do not believe we can rely on it. The only true way to view the lineage is the recursive approach being used.

This is correct that we cannot assume the history is ordered. I don't know what you mean by "the recursive approach". The current implementation is still incorrect due to the same reason - history may not be strictly ordered.

**Comment by xndai** [comment]
I open an issue here - https://github.com/apache/iceberg/issues/17358
